src/python/pants/backend/jvm/tasks/classpath_products.py

```diff
@@ -247,23 +249,41 @@ def get_for_targets(self, targets):
               jars.
     :rtype: list of (string, string)
     """
-    cp_entries = self.get_classpath_entries_for_targets(targets)
+    cp_entries = self.get_classpath_entries_for_targets(targets, context=context)
     return [(conf, cp_entry.path) for conf, cp_entry in cp_entries]
 
-  def get_classpath_entries_for_targets(self, targets, respect_excludes=True):
+  def get_classpath_entries_for_targets(self, targets, respect_excludes=True, context=None):
     """Gets the classpath products for the given targets.
 
     Products are returned in order, optionally respecting target excludes.
 
     :param targets: The targets to lookup classpath products for.
     :param bool respect_excludes: `True` to respect excludes; `False` to ignore them.
+    :param pants.goal.context.Context context: If not `None`, will attempt to materialize any
+      classpath entries to disk which were generated by hermetic execution and which have not
+      already been materialized to disk.
     :returns: The ordered (conf, classpath entry) tuples.
     :rtype: list of (string, :class:`ClasspathEntry`)
     """
 
     # remove the duplicate, preserve the ordering.
-    return list(OrderedSet([cp for cp, target in self.get_product_target_mappings_for_targets(
-                            targets, respect_excludes)]))
+    classpath_entries = list(OrderedSet([
+      cp for cp, target in self.get_product_target_mappings_for_targets(targets, respect_excludes)
+    ]))
+
+    if context is not None:
+      missing_entries = [
+        entry for (scope, entry) in classpath_entries
+        if entry.directory_digest and not entry.has_been_materialized
+      ]
+      if missing_entries:
```

Comment:
Is it correct that the `.create_new()` and `std::io::ErrorKind::AlreadyExists` check in `store.rs` ensures that it's safe to have multiple threads materializing potentially overlapping files at the same time here? I believe that is what the PR description says, but wanted to make sure.

Reply:
With this check, we avoid calling into the rust code at all; having to read the (transitive) directory structure out of the lmdb cache (on disk), do a bunch of `mkdirs`, make an `open` syscall per file, and handle the errors, is noticeably more overhead (particularly on non-SSDs) than just checking a boolean under a lock.